src/collapse/dtrace.rs

```diff
@@ -0,0 +1,214 @@
+use super::Collapse;
+use lazy_static::lazy_static;
+use regex::Regex;
+use std::collections::{HashMap, VecDeque};
+use std::io;
+use std::io::prelude::*;
+
+const HEADER_LINES: u8 = 3;
+
+lazy_static! {
+    static ref CPP_REGEX: Regex = Regex::new("(?P<n>::.*)[(<].*").unwrap();
+}
+
+/// Settings that change how frames are named from the incoming stack traces.
+///
+/// All options default to off.
+#[derive(Clone, Debug, Default)]
+pub struct Options {
+    /// include function offset (except leafs)
+    pub includeoffset: bool,
+}
+
+/// A stack collapser for the output of dtrace `ustrace()`.
+///
+/// To construct one, either use `dtrace::Folder::default()` or create an [`Options`] and use
+/// `dtrace::Folder::from(options)`.
+#[derive(Default)]
+pub struct Folder {
+    /// Function entries on the stack in this entry thus far.
+    stack: VecDeque<String>,
+
+    /// Number of times each call stack has been seen.
+    occurrences: HashMap<String, usize>,
+
+    /// Keep track of stack string size while we consume a stack
+    stack_str_size: usize,
+
+    opt: Options,
+}
+
+impl Collapse for Folder {
+    fn collapse<R, W>(&mut self, mut reader: R, writer: W) -> io::Result<()>
+    where
+        R: BufRead,
+        W: Write,
+    {
+        let mut line = String::new();
+        for _ in 0..HEADER_LINES {
+            reader.read_line(&mut line)?;
+        }
+
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            let line = line.trim();
+
+            if line.is_empty() {
+                continue;
+            } else if let Ok(count) = line.parse::<usize>() {
+                self.on_stack_end(count);
+            } else {
+                self.on_stack_line(line);
+            }
+        }
+        self.finish(writer)
+    }
+}
+
+impl From<Options> for Folder {
+    fn from(opt: Options) -> Self {
+        Self {
+            stack: VecDeque::default(),
+            occurrences: HashMap::default(),
+            opt: opt,
+            stack_str_size: 0,
```

Comment:
The perf collapse code just [computes this value](https://github.com/jonhoo/inferno/blob/10a1fd7c4c891dbc6a3c6ed74541d884eb89b1e9/src/collapse/perf.rs#L333); is there a particular reason you wanted to keep a running tally instead?

Reply:
I have the impression that performance matters to the project, so keeping a tally is cheaper then re-iterating over something we already went over during parsing.

Reply:
Sure, that's a fair point. I doubt that particular counting is very expensive since it's just `O(#stack segments)`, but seems fine to me to do it this way too :)